### 1.Dataset

A dataset (`custumer.csv`) is a structured collection of data containing multiple observations or records and their corresponding characteristics.

Real-world example:

A telecom company may maintain customer information in `custumer.csv` such as:

* Age
* Contract type
* Monthly charges
* Internet service
* Payment method
* Customer churn

Each customer represents one record, while each characteristic represents a feature.

In [4]:
df.head(10)

,CustomerID,Age,Gender,TenureYears,MonthlyCharges,TotalCharges,ContractType,InternetService,PaymentMethod,SeniorCitizen,SupportCalls,StreamingService,PaperlessBilling,Churn
0,CUST00001,56.0,Female,0,20.41,131.92,One year,DSL,Bank transfer,No,3,No,Yes,Yes
1,CUST00002,69.0,Male,9,21.10,2394.45,Two year,DSL,Credit card,Yes,1,Yes,No,No
2,CUST00003,46.0,Female,3,61.66,2755.45,Two year,DSL,Bank transfer,No,3,Yes,No,No
3,CUST00004,32.0,Male,8,35.50,3796.58,Month-to-month,No,Credit card,No,2,No,No,No
4,CUST00005,60.0,Male,9,86.45,10188.80,Two year,Fiber optic,Electronic check,Yes,3,No,Yes,Yes
5,CUST00006,25.0,Male,5,90.88,5818.04,Two year,DSL,Mailed check,No,0,No,No,No
6,CUST00007,38.0,Female,5,37.79,2670.55,Month-to-month,DSL,Mailed check,No,1,Yes,Yes,No
7,CUST00008,56.0,Male,0,108.85,1106.71,Two year,No,Electronic check,No,4,Yes,No,Yes
8,CUST00009,36.0,Male,9,64.19,7652.76,One year,DSL,Mailed check,No,3,Yes,No,No
9,CUST00010,40.0,Male,1,34.28,618.59,One year,Fiber optic,Electronic check,No,3,Yes,Yes,No


### 2.Feature

A feature is an individual variable or column in `custumer.csv` that describes some characteristic of an observation.

For example:

* Age
* MonthlyCharges
* ContractType
* InternetService
* PaymentMethod

These are all features of a customer.

Real-world example:

For predicting whether a customer will leave a telecom company:

* Age → feature
* Monthly charges → feature
* Contract type → feature
* Internet service → feature

In [5]:
features = df.columns

print(features)

Index(['CustomerID', 'Age', 'Gender', 'TenureYears', 'MonthlyCharges',
       'TotalCharges', 'ContractType', 'InternetService', 'PaymentMethod',
       'SeniorCitizen', 'SupportCalls', 'StreamingService', 'PaperlessBilling',
       'Churn'],
      dtype='object')


### 3.Target

The target is the variable that we want to predict or explain.

In our `custumer.csv` dataset:

Churn is the target.

Real-world example:

A telecom company wants to predict:

"Will this customer leave the company?"

Therefore:

Target = Churn

In [6]:
target = "Churn"

print(df[target].head())

0    Yes
1     No
2     No
3     No
4    Yes
Name: Churn, dtype: object


### 4. Independent Variables

Independent variables are variables used as inputs to predict the target.

They are also called:

* Input variables
* Predictors
* Features

For our `custumer.csv` dataset:

* Age
* Gender
* TenureYears
* MonthlyCharges
* TotalCharges
* ContractType
* InternetService
* PaymentMethod

are independent variables.

Real-world example:

We could use: Age + ContractType + MonthlyCharges + TenureYears

to predict: Churn

In [8]:
X = df.drop("Churn", axis=1)

print(X.head())

  CustomerID   Age  Gender  TenureYears  MonthlyCharges  TotalCharges  \
0  CUST00001  56.0  Female            0           20.41        131.92   
1  CUST00002  69.0    Male            9           21.10       2394.45   
2  CUST00003  46.0  Female            3           61.66       2755.45   
3  CUST00004  32.0    Male            8           35.50       3796.58   
4  CUST00005  60.0    Male            9           86.45      10188.80   

     ContractType InternetService     PaymentMethod SeniorCitizen  \
0        One year             DSL     Bank transfer            No   
1        Two year             DSL       Credit card           Yes   
2        Two year             DSL     Bank transfer            No   
3  Month-to-month              No       Credit card            No   
4        Two year     Fiber optic  Electronic check           Yes   

   SupportCalls StreamingService PaperlessBilling  
0             3               No              Yes  
1             1              Yes          

### 5. Dependent Variable

A dependent variable is the variable whose value depends on the independent variables.

In machine learning, this is generally the target variable.

For our `custumer.csv` dataset:

Dependent Variable = Churn

In [9]:
y = df["Churn"]

print(y.head())

0    Yes
1     No
2     No
3     No
4    Yes
Name: Churn, dtype: object


### 6. Numerical Features

Numerical features contain numbers and can be used for mathematical calculations.

Examples from our `custumer.csv` dataset:

* Age
* TenureYears
* MonthlyCharges
* TotalCharges
* SupportCalls

Real-world example:

MonthlyCharges could contain:

45.50,
72.30,
99.99

We can calculate:

* Average
* Minimum
* Maximum
* Standard deviation

In [11]:
numerical_columns = df.select_dtypes(include="number").columns

print(numerical_columns)

Index(['Age', 'TenureYears', 'MonthlyCharges', 'TotalCharges', 'SupportCalls'], dtype='object')


In [12]:
df[numerical_columns].describe()

,Age,TenureYears,MonthlyCharges,TotalCharges,SupportCalls
count,4969.000000,5020.000000,4970.000000,4969.000000,5020.000000
mean,43.328034,4.911753,69.480559,4126.251849,2.105777
std,14.762282,3.137479,28.678515,3237.099088,1.630602
min,18.000000,0.000000,20.020000,72.450000,0.000000
25%,31.000000,2.000000,44.602500,1440.880000,1.000000
50%,43.000000,5.000000,69.285000,3355.920000,2.000000
75%,56.000000,8.000000,93.770000,6068.800000,3.000000
max,69.000000,10.000000,119.970000,14603.300000,9.000000


### 7. Categorical Features

Categorical features contain categories or labels instead of continuous numerical measurements.

Examples in `custumer.csv`:

* Gender
* ContractType
* InternetService
* PaymentMethod
* StreamingService
* PaperlessBilling

Real-world example:

ContractType can have:

* Month-to-month
* One year
* Two year

These values represent categories.

In [14]:
categorical_columns = df.select_dtypes(include="object").columns

print(categorical_columns)

Index(['CustomerID', 'Gender', 'ContractType', 'InternetService',
       'PaymentMethod', 'SeniorCitizen', 'StreamingService',
       'PaperlessBilling', 'Churn'],
      dtype='object')


### 8. Discrete Features

Discrete features contain countable values, usually whole numbers.

Examples:

* Number of support calls
* Number of purchases
* Number of children
* Number of transactions

In our `custumer.csv` dataset:

SupportCalls is a discrete numerical feature.

Example:
0

1

2

3

4

5

You cannot normally have:

2.73 support calls

In [16]:
print(df["SupportCalls"].head(10))

0    3
1    1
2    3
3    2
4    3
5    0
6    1
7    4
8    3
9    3
Name: SupportCalls, dtype: int64


### 9. Continuous Features

Continuous features can take a range of numerical values, including decimal values.

Examples:

* Height
* Weight
* Temperature
* Salary
* MonthlyCharges

In our `custumer.csv` dataset:

* MonthlyCharges
* TotalCharges

are continuous features.

Example:

49.99

72.45

89.75

103.20

In [17]:
print(df["MonthlyCharges"].head(10))

0     20.41
1     21.10
2     61.66
3     35.50
4     86.45
5     90.88
6     37.79
7    108.85
8     64.19
9     34.28
Name: MonthlyCharges, dtype: float64


### 10. Identifier Columns

An identifier column uniquely identifies each record.

In our `custumer.csv` dataset:

CustomerID is an identifier.

Example:

CUST00001

CUST00002

CUST00003

#### Important ML concept:

An identifier generally should not be used as a predictive feature because it doesn't represent meaningful customer behavior.

For example:

CustomerID = CUST00001

doesn't tell us whether the customer will churn.

In [18]:
print(df["CustomerID"].head())

0    CUST00001
1    CUST00002
2    CUST00003
3    CUST00004
4    CUST00005
Name: CustomerID, dtype: object


### 11. Target Variable

The target variable is the output that a machine learning model tries to predict.

Our target in `custumer.csv`:

Churn contains:

* Yes
* No

This is a classification problem because the target contains categories.

Real-world example:

A telecom company could use the model to identify customers who are likely to leave.

Customer information

        ↓
        
Machine Learning Model

        ↓
        
Churn prediction

        ↓
        
Yes / No

In [19]:
df["Churn"].value_counts()

Churn
No     3565
Yes    1455
Name: count, dtype: int64

In [20]:
df["Churn"].value_counts(normalize=True) * 100

Churn
No     71.015936
Yes    28.984064
Name: proportion, dtype: float64

### 12. Metadata

Metadata means information that describes the `custumer.csv` dataset and its columns.

Examples:

* Dataset name (`custumer.csv`)
* Number of rows
* Number of columns
* Column names
* Data types
* Missing values
* Description of columns
* Target variable
* Data source

In [21]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5020 entries, 0 to 5019
Data columns (total 14 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   CustomerID        5020 non-null   object 
 1   Age               4969 non-null   float64
 2   Gender            5020 non-null   object 
 3   TenureYears       5020 non-null   int64  
 4   MonthlyCharges    4970 non-null   float64
 5   TotalCharges      4969 non-null   float64
 6   ContractType      5020 non-null   object 
 7   InternetService   5020 non-null   object 
 8   PaymentMethod     5020 non-null   object 
 9   SeniorCitizen     5020 non-null   object 
 10  SupportCalls      5020 non-null   int64  
 11  StreamingService  5020 non-null   object 
 12  PaperlessBilling  5020 non-null   object 
 13  Churn             5020 non-null   object 
dtypes: float64(3), int64(2), object(9)
memory usage: 549.2+ KB


### 13. Dataset Dimensions

Dataset dimensions describe the number of rows and columns in `custumer.csv`.

For example:

Rows = 5020

Columns = 14

In [22]:
df.shape

(5020, 14)

In [23]:
rows, columns = df.shape

print("Number of rows in custumer.csv:", rows)
print("Number of columns in custumer.csv:", columns)

Number of rows in custumer.csv: 5020
Number of columns in custumer.csv: 14
